R/group2NA.R

```diff
@@ -21,54 +21,67 @@
 # elong <- tidyr::gather(economics, variable, value, -date)
 # plot_ly(group2NA(elong, "variable"), x = ~date, y = ~value)
 # 
+# @import data.table
+#
 
 group2NA <- function(data, groupNames = "group", nested = NULL, ordered = NULL,
                      retrace.first = inherits(data, "GeomPolygon")) {
+  
   if (NROW(data) == 0) return(data)
-  data <- data[!duplicated(names(data))]
-  # a few workarounds since dplyr clobbers classes that we rely on in ggplotly
+  
+  # make copy via assignment and eliminate duplicated column names 
+  if(data.table::is.data.table(data)){
+    data <- data[,unique(names(data)),with=FALSE]
+  } else {
+    data <- data[!duplicated(names(data))]
+  }
+  
+  # store class information from function input
   retrace <- force(retrace.first)
   datClass <- class(data)
   
   # sanitize variable names
   groupNames <- groupNames[groupNames %in% names(data)]
   nested <- nested[nested %in% names(data)]
   ordered <- ordered[ordered %in% names(data)]
-  # ignore any already existing groups
-  data <- dplyr::ungroup(data)
   
   # if group doesn't exist, just arrange before returning
   if (!length(groupNames)) {
     if (length(ordered)) {
-      data <- dplyr::arrange_(data, c(nested, ordered))
+      return(
+          data.table::setDT(data,key = c(nested, ordered)) %>%
+          structure(class = datClass)
+      )
+    } else {
+      return(data)
     }
-    return(data)
   }
+  
   allVars <- c(nested, groupNames, ordered)
-  for (i in allVars) {
-    data <- dplyr::group_by_(data, i, add = TRUE)
-  }
-  # first, arrange everything
-  data <- dplyr::do(data, dplyr::arrange_(., allVars))
-  data <- dplyr::ungroup(data)
-  for (i in c(nested, groupNames)) {
-    data <- dplyr::group_by_(data, i, add = TRUE)
-  }
-  # TODO: this is slow, can it be done with dplyr::slice()?
-  d <- if (retrace.first) {
-    dplyr::do(data, rbind.data.frame(., .[1,], NA))
+  
+  # if retrace.first is TRUE,repeat the first row of each group and add an empty row of NA's after each group.
+  # if retrace.first is FALSE, just add an empty row to each group.
+  # delete final row of NA's, return d with the original class
+  
+  if (retrace.first) {
+    return(
+      data.table::setDT(data, key = allVars)[ data[, .I[c(seq_along(.I), 1L, .N+1L)], by=allVars]$V1 ][-.N,] %>% 
+        structure(class = datClass)
+    )
   } else {
-    dplyr::do(data, rbind.data.frame(., NA))
+    return(      
+        data.table::setDT(data, key = allVars)[ data[, .I[c(seq_along(.I), .N+1L)], by=allVars]$V1 ][-.N,] %>%
+        structure(class = datClass)
```

Comment:
@msummersgill this isn't quite right. We don't want to separate groups (aka insert NAs) according to `allVars`. Instead, only groups defined by `groupNames` should be separated (after the data has been sorted appropriately).